File: sdk/swift/Sources/Gosslens/MicInput.swift
import AVFoundation

/// Feeds the device microphone into the session: the web SDK's GossMicInput, brought here so
/// practice, live and capture do not each write their own interleave. The engine resamples
/// whatever rate the device hands over, so this passes the hardware format through.
public final class GossMicInput {
    private let session: GossSession
    private let engineHandle: GossEngine
    private let engine = AVAudioEngine()
    private var interleaved: [Float] = []
    private var running = false

    public init(session: GossSession, engine: GossEngine) {
        self.session = session
        self.engineHandle = engine
    }

    /// Taps the input node and submits every buffer. Requires the app to have arranged microphone
    /// permission and an active, category-appropriate `AVAudioSession`; this deliberately does not
    /// touch either, because the app owns that policy and a call or broadcast may already hold it.
    public func start() throws {
        guard !running else { return }
        let input = engine.inputNode
        let format = input.inputFormat(forBus: 0)
        guard format.sampleRate > 0, format.channelCount > 0 else { throw GossStatus.invalidArgument }
        input.installTap(onBus: 0, bufferSize: 2048, format: format) { [weak self] buffer, when in
            self?.submit(buffer, at: when)
        }
        engine.prepare()
        try engine.start()
        running = true
    }

    public func stop() {
        guard running else { return }
        engine.inputNode.removeTap(onBus: 0)
        engine.stop()
        running = false
    }

    /// Interleaves the platform's per-channel float buffer and hands it over. Reuses one array so
    /// a tap running every few milliseconds allocates nothing after the first buffer.
    private func submit(_ buffer: AVAudioPCMBuffer, at when: AVAudioTime) {
        guard let channels = buffer.floatChannelData else { return }
        let frames = Int(buffer.frameLength)
        let count = Int(buffer.format.channelCount)
        guard frames > 0, count > 0 else { return }
        let needed = frames * count
        if interleaved.count < needed { interleaved = [Float](repeating: 0, count: needed) }
        interleaved.withUnsafeMutableBufferPointer { out in
            for c in 0..<count {
                let src = channels[c]
                for f in 0..<frames { out[f * count + c] = src[f] }
            }
        }
        let seconds = Double(when.sampleTime) / when.sampleRate
        let stamp = Int64(seconds * 1_000_000)
        try? engineHandle.submitAudio(session: session, samples: Array(interleaved[0..<needed]),
                                frameCount: UInt32(frames), sampleRate: UInt32(buffer.format.sampleRate),
                                channels: UInt32(count), timestampUs: stamp)
    }
}

File: sdk/kotlin/src/main/kotlin/com/gosslens/MicInput.kt
package com.gosslens

import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Feeds the device microphone into the session: the web SDK's GossMicInput, brought here so
 * practice, live and capture do not each write their own float conversion. The engine resamples
 * whatever rate the device grants, so this submits the granted rate rather than the requested one.
 */
class GossMicInput(
    private val engine: GossEngine,
    private val session: GossSession,
    private val sampleRate: Int = 48_000,
) {
    private var record: AudioRecord? = null
    private var reader: Thread? = null
    @Volatile private var running = false

    /**
     * Starts the capture thread. The caller must already hold RECORD_AUDIO; this deliberately does
     * not request it, because the app owns that policy and a call may already hold the microphone.
     * Returns false when the platform refuses to open a recorder.
     */
    fun start(): Boolean {
        if (running) return true
        val minBytes = AudioRecord.getMinBufferSize(
            sampleRate,
            AudioFormat.CHANNEL_IN_MONO,
            AudioFormat.ENCODING_PCM_FLOAT,
        )
        if (minBytes <= 0) return false
        val rec = try {
            AudioRecord(
                MediaRecorder.AudioSource.VOICE_COMMUNICATION,
                sampleRate,
                AudioFormat.CHANNEL_IN_MONO,
                AudioFormat.ENCODING_PCM_FLOAT,
                minBytes * 2,
            )
        } catch (_: SecurityException) {
            return false
        }
        if (rec.state != AudioRecord.STATE_INITIALIZED) {
            rec.release()
            return false
        }
        // The rate the platform granted, which is not always the one asked for.
        val granted = rec.sampleRate
        val frames = FRAMES_PER_READ
        // One direct buffer for the life of the capture, so a read every few milliseconds
        // allocates nothing.
        val buffer = ByteBuffer.allocateDirect(frames * 4).order(ByteOrder.nativeOrder())
        val floats = FloatArray(frames)
        rec.startRecording()
        record = rec
        running = true
        reader = Thread {
            var at = 0L
            while (running) {
                val read = rec.read(floats, 0, frames, AudioRecord.READ_BLOCKING)
                if (read <= 0) continue
                buffer.clear()
                buffer.asFloatBuffer().put(floats, 0, read)
                val stamp = at * 1_000_000L / granted
                engine.submitAudio(session, buffer, read, granted, 1, stamp)
                at += read
            }
        }.also { it.isDaemon = true; it.start() }
        return true
    }

    fun stop() {
        if (!running) return
        running = false
        reader?.join(500)
        reader = null
        record?.let {
            it.stop()
            it.release()
        }
        record = null
    }

    private companion object {
        const val FRAMES_PER_READ = 1024
    }
}
